File: aoc2025/javascript/day1.js
import { readFile } from './utilities.js';

class DialInstruction {
  constructor(direction, step) {
    this.direction = direction;
    this.step = step;
  }
}

class Dial {
  constructor(value, next = null, prev = null) {
    this.value = value;
    this.next = next;
    this.prev = prev;
  }
}

class DialList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  next() {
    this.head = this.head.next;
  }

  prev() {
    this.head = this.head.prev;
  }

  buildLinkedList() {
    let prevNode = null;
    for (let i = 0; i < 100; i++) {
      const node = new Dial(i, null, prevNode);

      if (this.head === null) {
        this.head = node;
      }

      if (prevNode !== null) {
        prevNode.next = node;
      }

      this.tail = node;
      prevNode = node;
    }

    this.tail.next = this.head;
    this.head.prev = this.tail;
  }
}

function parseInstructions(lines) {
  const instructions = [];
  for (const line of lines) {
    const direction = line[0];
    const step = parseInt(line.substring(1));
    instructions.push(new DialInstruction(direction, step));
  }
  return instructions;
}

function day1() {
  const start = performance.now();

  const files = readFile('day1.part1');
  const dialInstructions = parseInstructions(files);

  const dialList = new DialList();
  dialList.buildLinkedList();

  while (dialList.head.value !== 50) {
    dialList.next();
  }

  let day1part1Count = 0;
  let day1part2Count = 0;

  for (const dialInstruction of dialInstructions) {
    switch (dialInstruction.direction) {
      case 'R':
        for (let j = 0; j < dialInstruction.step; j++) {
          dialList.next();
          if (dialList.head.value === 0) {
            day1part2Count++;
          }
        }
        break;
      case 'L':
        for (let j = 0; j < dialInstruction.step; j++) {
          dialList.prev();
          if (dialList.head.value === 0) {
            day1part2Count++;
          }
        }
        break;
    }
    if (dialList.head.value === 0) {
      day1part1Count++;
    }
  }

  const elapsed = ((performance.now() - start) / 1000).toFixed(4);
  console.log(`Day 1:`);
  console.log(`  Part 1: ${day1part1Count}, Part 2: ${day1part2Count} (${elapsed}s)`);
}

function day1_1() {
  const start = performance.now();

  const files = readFile('day1.part1');
  const dialInstructions = parseInstructions(files);

  let step1Count = 0;
  let step2Count = 0;
  let currentValue = 50;

  for (const dialInstruction of dialInstructions) {
    switch (dialInstruction.direction) {
      case 'R': {
        const newValue = currentValue + dialInstruction.step;
        if (newValue >= 100) {
          if (currentValue === 0) {
            step2Count += Math.floor(dialInstruction.step / 100);
          } else {
            step2Count += Math.floor((dialInstruction.step + currentValue) / 100);
          }
        }
        currentValue = newValue % 100;
        break;
      }
      case 'L': {
        const newValue = currentValue - dialInstruction.step;
        if (currentValue === 0) {
          step2Count += Math.floor(dialInstruction.step / 100);
        } else if (dialInstruction.step >= currentValue) {
          step2Count += Math.floor((dialInstruction.step - currentValue) / 100) + 1;
        }
        currentValue = ((newValue % 100) + 100) % 100;
        break;
      }
    }
    if (currentValue === 0) {
      step1Count++;
    }
  }

  const elapsed = ((performance.now() - start) / 1000).toFixed(4);
  console.log(`\nDay 1 Optimized:`);
  console.log(`  Part 1: ${step1Count}, Part 2: ${step2Count} (${elapsed}s)`);
}

day1();
day1_1();

File: aoc2025/python/day1.py
import time
from utilities import read_file


class Dial:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None


class DialList:
    def __init__(self):
        self.head = None
        self.tail = None

    def next(self):
        self.head = self.head.next

    def prev(self):
        self.head = self.head.prev

    def build_linked_list(self):
        prev_node = None
        for i in range(100):
            node = Dial(i)
            node.prev = prev_node

            if self.head is None:
                self.head = node

            if prev_node is not None:
                prev_node.next = node

            self.tail = node
            prev_node = node

        self.tail.next = self.head
        self.head.prev = self.tail


def parse_instructions(lines):
    instructions = []
    for line in lines:
        direction = line[0]
        step = int(line[1:])
        instructions.append((direction, step))
    return instructions


def day1():
    start_time = time.time()

    lines = read_file("day1.part1")
    instructions = parse_instructions(lines)

    dial_list = DialList()
    dial_list.build_linked_list()

    while dial_list.head.value != 50:
        dial_list.next()

    part1_count = 0
    part2_count = 0

    for direction, step in instructions:
        if direction == 'R':
            for _ in range(step):
                dial_list.next()
                if dial_list.head.value == 0:
                    part2_count += 1
        elif direction == 'L':
            for _ in range(step):
                dial_list.prev()
                if dial_list.head.value == 0:
                    part2_count += 1

        if dial_list.head.value == 0:
            part1_count += 1

    elapsed = time.time() - start_time
    print(f"Day 1: part1={part1_count}, part2={part2_count}, time={elapsed:.4f}s")


def day1_optimized():
    start_time = time.time()

    lines = read_file("day1.part1")
    instructions = parse_instructions(lines)

    part1_count = 0
    part2_count = 0
    current_value = 50

    for direction, step in instructions:
        if direction == 'R':
            new_value = current_value + step
            if new_value >= 100:
                if current_value == 0:
                    part2_count += step // 100
                else:
                    part2_count += (step + current_value) // 100
            current_value = new_value % 100
        elif direction == 'L':
            new_value = current_value - step
            if current_value == 0:
                part2_count += step // 100
            elif step >= current_value:
                part2_count += (step - current_value) // 100 + 1
            current_value = ((new_value % 100) + 100) % 100

        if current_value == 0:
            part1_count += 1

    elapsed = time.time() - start_time
    print(f"Day 1 (optimized): part1={part1_count}, part2={part2_count}, time={elapsed:.4f}s")


if __name__ == "__main__":
    day1()
    day1_optimized()
